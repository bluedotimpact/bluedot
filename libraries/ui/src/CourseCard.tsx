import React from 'react';
import clsx from 'clsx';
import { Card } from './Card';
import { CTALinkOrButton } from './CTALinkOrButton';
import { Tag } from './Tag';

export type CourseCardProps = React.PropsWithChildren<{
  title: string,
  description: string,
  imageSrc: string,
  className?: string,
  ctaUrl?: string,
  applicationDeadline?: string, // Expected format: "Feb 1"
  // eslint-disable-next-line react/no-unused-prop-types
  cardType?: 'Featured' | 'Regular',
  // eslint-disable-next-line react/no-unused-prop-types
  courseType?: CourseType,
}>;

type CourseType = 'Crash course' | 'Self-paced' | 'In-depth course';

const courseLength = (courseType: CourseType) => {
  switch (courseType) {
    case 'Crash course':
      return '5 days';
    case 'Self-paced':
      return '12 weeks';
    case 'In-depth course':
      return '12 weeks';
    default:
      throw new Error(`Invalid course type: ${courseType satisfies never}`);
  }
};

const applyByText = (applicationDeadline: string | undefined) => {
  return applicationDeadline ? `Apply by ${applicationDeadline}` : 'Apply now';
};

const FeaturedCourseCard: React.FC<CourseCardProps> = ({
  className,
  title,
  description,
  ctaUrl,
  applicationDeadline,
  imageSrc,
  children,
}) => {
  const wrapperClassName = clsx(
    'course-card course-card--featured card flex flex-col items-start transition-transform duration-200 hover:scale-[1.01] hover:container-elevated',
    'container-lined p-6 max-w-[656px]',
    className,
  );

  return (
    <a
      href={ctaUrl}
      className={wrapperClassName}
    >
      <div className="course-card__content flex gap-space-between">
        <div className="course-card__text">
          <p className="course-card__featured-label uppercase font-[650] text-xs mb-3">
            Featured course
          </p>
          <h3 className="course-card__title mb-6">
            {title}
          </h3>
          <CTALinkOrButton
            className="course-card__cta mb-6 px-6"
            variant="primary"
            withChevron={false}
          >
            {applyByText(applicationDeadline)}
          </CTALinkOrButton>
          {description && (
            <p className="course-card__subtitle text-sm overflow-hidden text-bluedot-black text-ellipsis mb-6">
              {description}
            </p>
          )}
        </div>
        <div className="course-card__image-container flex-shrink-0 max-w-[60%] mb-6">
          <img
            className="course-card__image w-[319px] h-full object-cover"
            src={imageSrc}
            alt={`${title}`}
          />
        </div>
      </div>
      {children}
    </a>
  );
};

export const CourseCard: React.FC<CourseCardProps> = ({
  className,
  title,
  description,
  ctaUrl,
  applicationDeadline,
  cardType = 'Regular',
  courseType = 'Crash course',
  imageSrc,
}) => {
  const CourseCardFooter = (
    <div className="course-card__footer flex justify-between w-full">
      <p className="course-card__footer-left text-left text-xs text-bluedot-black">
        <span className="course-card__length font-medium">{courseLength(courseType)}</span>
      </p>
      <Tag className="course-card__type">{courseType}</Tag>
    </div>
  );

  return cardType === 'Featured' ? (
    <div className="p-[3px]">
      <FeaturedCourseCard
        className={className}
        title={title}
        description={description}
        ctaUrl={ctaUrl}
        applicationDeadline={applicationDeadline}
        imageSrc={imageSrc}
      >
        {CourseCardFooter}
      </FeaturedCourseCard>
    </div>
  ) : (
    <div className="p-[2px]">
      <Card
        imageSrc={imageSrc}
        title={title}
        subtitle={description}
        ctaUrl={ctaUrl}
        isEntireCardClickable
        isExternalUrl
        className={clsx(
          'course-card course-card--regular container-lined p-5',
          'flex flex-col w-[323px] h-[466px] hover:container-elevated',
          className,
        )}
        imageClassName="course-card__image w-full h-[165px] object-cover rounded-none"
        subtitleClassName="flex-grow overflow-hidden text-ellipsis line-clamp-4 max-h-[96px]"
      >
        {CourseCardFooter}
      </Card>
    </div>
  );
};
