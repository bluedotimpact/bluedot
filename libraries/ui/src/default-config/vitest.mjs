/* eslint-disable import/no-extraneous-dependencies */
import { defineConfig, defaultExclude } from 'vitest/config';
import react from '@vitejs/plugin-react';
import * as dotenv from 'dotenv';

/**
 * @param {import('vitest/config').UserConfig} [config] 
 * @returns {import('vitest/config').UserConfig}
 */
export const withDefaultBlueDotVitestConfig = async (config) => defineConfig({
  ...config,
  plugins: [react(), ...(config?.plugins ?? [])],
  test: {
    environment: 'happy-dom',
    env: dotenv.config({ path: '.env.test' }).parsed,

    ...config?.test,

    exclude: [...defaultExclude, '**/.{turbo,next}/**', ...(config?.test?.exclude ?? [])],
  },
});
