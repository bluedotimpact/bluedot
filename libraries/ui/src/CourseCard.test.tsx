import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import { CourseCard } from './CourseCard';

describe('CourseCard', () => {
  const baseProps = {
    title: 'Introduction to Testing',
    description: 'Learn proper testing techniques',
    courseType: 'Crash course',
    image: '/test-image.jpg',
  };

  test('renders default as expected', () => {
    const { container } = render(<CourseCard {...baseProps} />);
    expect(container).toMatchSnapshot();
  });

  test('renders with optional args', () => {
    const { container } = render(
      <CourseCard
        {...baseProps}
        className="custom-class"
        cardType="Featured"
      />,
    );
    expect(container).toMatchSnapshot();
  });

  test('renders with optional yield', () => {
    const { container } = render(
      <CourseCard {...baseProps}>
        <button test-id="test-element" type="button">Enroll Now</button>
      </CourseCard>,
    );
    const yieldedContent = screen.getByTestId('test-element');
    expect(yieldedContent).toBeTruthy();
  });
});
