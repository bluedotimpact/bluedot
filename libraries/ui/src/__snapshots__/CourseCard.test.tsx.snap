// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CourseCard > renders Featured as expected 1`] = `
<div>
  <a
    class="course-card course-card--featured card flex flex-col items-start justify-between m-[3px] container-lined p-6 max-w-[700px] transition-transform duration-200 hover:scale-[1.01] hover:container-elevated"
    href="https://bluedot.org/courses/what-the-fish"
  >
    <div
      class="course-card__content block md:flex gap-space-between w-full"
    >
      <div
        class="course-card__text"
      >
        <p
          class="course-card__featured-label uppercase font-[650] text-size-xs mb-3 bluedot-p"
        >
          Featured course
        </p>
        <h3
          class="course-card__title mb-6 bluedot-h3"
        >
          Title
        </h3>
        <button
          class="cta-button flex items-center justify-center rounded-sm transition-all duration-200 text-sm px-4 py-3 w-fit font-[650] whitespace-nowrap cursor-pointer cta-button--primary bg-bluedot-normal link-on-dark course-card__cta mb-6 px-6"
          data-testid="cta-button"
          type="button"
        >
          <span
            class="cta-button__text"
          >
            Apply now
          </span>
        </button>
        <p
          class="course-card__subtitle overflow-hidden text-bluedot-black text-ellipsis mb-6 bluedot-p"
        >
          Description
        </p>
      </div>
      <div
        class="course-card__image-container shrink-0 md:max-w-[60%] w-fit mb-6"
      >
        <img
          alt=""
          class="course-card__image size-full object-cover"
          src="/images/courses/course.jpg"
        />
      </div>
    </div>
    <div
      class="course-card__footer flex justify-between w-full"
    >
      <p
        class="course-card__footer-left text-left text-size-xs text-bluedot-black bluedot-p"
      >
        <span
          class="course-card__length font-medium"
        />
      </p>
      <span
        class="tag container-lined inline-flex items-center px-4 py-2 text-xs font-semibold text-color-secondary-text course-card__type"
        role="status"
      />
    </div>
  </a>
</div>
`;

exports[`CourseCard > renders default as expected 1`] = `
<div>
  <div
    class="p-[3px] h-full"
  >
    <a
      class="card flex flex-col items-start transition-transform duration-200 hover:scale-[1.01] course-card course-card--regular container-lined p-5 flex flex-col max-w-full h-full hover:container-elevated"
      href="https://bluedot.org/courses/what-the-fish"
    >
      <div
        class="card__image-container w-full mb-4"
      >
        <img
          alt="Title"
          class="card__image max-w-full max-h-full object-cover rounded-lg course-card__image w-full h-[165px] object-cover rounded-none"
          src="/images/courses/course.jpg"
        />
      </div>
      <div
        class="card__content flex flex-col gap-6 w-full flex-1 justify-between"
      >
        <div
          class="card__text"
        >
          <p
            class="card__title bluedot-h4 mb-2"
          >
            Title
          </p>
          <p
            class="card__subtitle bluedot-p grow overflow-hidden text-ellipsis line-clamp-4 max-h-[96px]"
          >
            Description
          </p>
        </div>
        <div
          class="card__bottom-section flex flex-col gap-space-between"
        >
          <div
            class="card__footer flex items-center justify-between w-full"
          >
            <div
              class="course-card__footer flex justify-between w-full"
            >
              <p
                class="course-card__footer-left text-left text-size-xs text-bluedot-black bluedot-p"
              >
                <span
                  class="course-card__length font-medium"
                />
              </p>
              <span
                class="tag container-lined inline-flex items-center px-4 py-2 text-xs font-semibold text-color-secondary-text course-card__type"
                role="status"
              />
            </div>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`CourseCard > renders with optional args 1`] = `
<span
  class="course-card__length font-medium"
/>
`;

exports[`CourseCard > renders with optional args 2`] = `
<img
  alt="Title"
  class="card__image max-w-full max-h-full object-cover rounded-lg course-card__image w-full h-[165px] object-cover rounded-none"
  src="/images/team/custom-size.jpg"
/>
`;
